task 1/main.cpp

```diff
@@ -0,0 +1,251 @@
+ * @author ZhenhuaZhuang
+ */
+#include <iostream>
+#include<vector>
+#include<opencv2/opencv.hpp>
+#include<opencv2/core/core.hpp>
+#include<opencv2/highgui/highgui.hpp>
+#include <opencv2/imgproc/types_c.h>
+using namespace std;
+using namespace cv;
+
+/* 参数定义 */
+#define _light_radio 3.2 //灯条长宽比最小值
+#define _light_length_min 80.0 //灯条长度最小值
+#define _light_length_max 140.0 //灯条长度最大值
+#define _light_angle_min 45.0 //灯条角度最小值
+#define _light_angle_max 135.0 //灯条角度最大值
+#define _armor_angleDiff_ 7 //装甲板左右两侧灯条角差
+#define _armor_LenDiff_ratio 0.2 //装甲板左右灯条长度差比
+#define _armor_yDiff_ratio_max 2.0 //装甲板左右灯条中心点y的差值比最大值
+#define _armor_xDiff_ratio_min 0.5 //装甲板左右灯条中心点x的差值比最小值
+#define _armor_radio_min 1.0 //装甲板左右灯条相距距离与灯条长度比值最小值
+#define _armor_radio_max 5.0 //装甲板左右灯条相距距离与灯条长度比值最大值
+
+/* 计算两点之间的距离 */
+double getDistance(const Point2f& pointO, const Point2f& point1)   
+{   
+    double distance;  
+    distance = powf((pointO.x - point1.x), 2) + powf((pointO.y - point1.y), 2);  
+    distance = sqrtf(distance);
+    return distance;
+}  
+
+/* 绘制旋转矩形 */
+void rotatedrectangle(const Mat& src, const RotatedRect& rec, const Scalar& c, int width)
+{
+   Point2f verticles[4];
+   rec.points(verticles);
+   for (int i = 0;i < 4;i++)
+      line(src, verticles[i], verticles[(i + 1) % 4], c, width);
+}
+
+/* 绘制装甲板 */
+void paint_armors(const Mat& src, const RotatedRect& leftrec, const RotatedRect& rightrec, const Scalar& c, int width)
+{  
+   Point2f center = (leftrec.center + rightrec.center) / 2;
+   float _length = getDistance(leftrec.center, rightrec.center);
+   float _width = _length * 0.8;
+   RotatedRect armor = RotatedRect(center, Point2f(_length, _width), leftrec.angle);
+   rotatedrectangle(src, armor, c, width);
+}
+
+int main()
+{
+    Mat srcImage, _grayImage1, _grayImage2, binImage1, binImage2, contoursImage, lightsImage, matchImage, armorsImage;
+    srcImage = imread("test.jpg");
+    if(!srcImage.data) {
+       cout << "图像加载失败!" << endl;
+       return 0;
+    }
+    else
+       cout << "图像加载成功!" << endl;
+    
+    contoursImage = srcImage.clone();
+    lightsImage = srcImage.clone();
+    armorsImage = srcImage.clone();
+    matchImage = srcImage.clone();
+
+    /* 进行BGR通道分离 */
+    vector<Mat> channels;
+    split(srcImage, channels);
+    _grayImage1 = channels[2] - channels[0];// Get red - blue image
+    _grayImage2 = channels[0];// Get blue image
+   
+    /* 阈值化 */
+    threshold(_grayImage1, binImage1, 218, 255, THRESH_BINARY);
+    threshold(_grayImage2, binImage2, 230, 255, THRESH_BINARY);
+
+    namedWindow("二值化", 0);
+    imshow("二值化", binImage1);
+    imwrite("二值化.jpg", binImage1);
+
+    Mat element1 = getStructuringElement(MORPH_ELLIPSE, Size(3, 3));
+    Mat element2 = getStructuringElement(MORPH_ELLIPSE, Size(4, 4));
+
+    /* 膨胀 */
+    dilate(binImage1, binImage1, element1);
+    dilate(binImage2, binImage2, element2);
+    
+    vector<RotatedRect> lightInfos1;
+    vector<RotatedRect> lightInfos2;
+
+    /* 找红色轮廓 */
```

Comment:
it would be redundant if you do the same job twice. Try to use some constant as your filtering threshold.